python/pylibcudf/pylibcudf/contiguous_split.pyx

```diff
@@ -111,6 +127,165 @@ cdef class PackedColumns:
+
+        Returns
+        -------
+        Size of packed data.
+        """
+        with nogil:
+            return dereference(self.c_obj).get_total_contiguous_size()
+
+    cpdef size_t next(self, DeviceBuffer buf):
+        """
+        Pack the next chunk into the provided device buffer.
+
+        Parameters
+        ----------
+        buf
+            The device buffer to pack into, must be the same size as
+            the `user_buffer_size` used to construct the packer.
+
+        Returns
+        -------
+        Number of bytes packed.
+        """
+        cdef device_span[uint8_t] d_span = device_span[uint8_t](
+            <uint8_t *>buf.c_data(), buf.c_size()
+        )
+        with nogil:
+            return dereference(self.c_obj).next(d_span)
+
+    cpdef memoryview build_metadata(self):
+        """
+        Build the metadata for the packed representation.
+
+        Returns
+        -------
+        memoryview of metadata suitable for passing to `unpack_from_memoryviews`.
+        """
+        cdef unique_ptr[vector[uint8_t]] metadata
+        with nogil:
+            metadata = move(dereference(self.c_obj).build_metadata())
+        return memoryview(HostBuffer.from_unique_ptr(move(metadata)))
+
+    cpdef tuple pack_to_host(self, DeviceBuffer buf):
+        """
+        Pack the entire table into a host buffer.
+
+        Parameters
+        ----------
+        buf
+           The device buffer to use as a staging buffer, must be the
+           same size as the `user_buffer_size` used to construct the
+           packer.
+
+        Returns
+        -------
+        tuple of metadata and packed host data (as memoryviews)
+
+        Notes
+        -----
+        This is stream-ordered with respect to the stream used when
+        creating the `ChunkedPack` and syncs that stream before returning.
+        """
+        cdef size_t offset = 0
+        cdef size_t size
+        cdef device_span[uint8_t] d_span = device_span[uint8_t](
+            <uint8_t *>buf.c_data(), buf.c_size()
+        )
+        cdef cudaError_t err
+        cdef unique_ptr[vector[uint8_t]] h_buf = (
+            make_unique[vector[uint8_t]](
+                dereference(self.c_obj).get_total_contiguous_size()
+            )
+        )
+        cdef cudaStream_t stream = self.stream.view().value()
+        with nogil:
+            while dereference(self.c_obj).has_next():
+                size = dereference(self.c_obj).next(d_span)
+                err = cudaMemcpyAsync(
+                    dereference(h_buf).data() + offset,
+                    d_span.data(),
+                    size,
+                    cudaMemcpyKind.cudaMemcpyDeviceToHost,
+                    stream,
+                )
+                offset += size
+                if err != cudaError.cudaSuccess:
+                    raise RuntimeError(
+                        f"Memcpy in pack_to_host failed error: {err}"
+                    )
+        self.stream.synchronize()
```

Comment:
Should we still sync the stream even after raising an exception? Without it, I think we might exit early while the stream still has pending memory operations.

Reply:
Good idea, done.